File: app-python/activities.py
import asyncio
import logging

from temporalio import activity
from temporalio.exceptions import ApplicationError

from dataclasses import dataclass

@dataclass
class BookTripInput:
    book_user_id: str
    book_car_id: str
    book_hotel_id: str
    book_flight_id: str

logging.basicConfig(level=logging.INFO)


@activity.defn
async def book_car(input: BookTripInput) -> str:
    activity.logger.info(f"Booking car {input.book_car_id}")

    await asyncio.sleep(1)

    if "flaky" in input.book_car_id.lower():
        # a transient error, which will be retried
        if activity.info().attempt < 6:
            raise ApplicationError("Car booking service is currently unavailable")

    return f"Booked car: {input.book_car_id}"


@activity.defn
async def book_hotel(input: BookTripInput) -> str:
    activity.logger.info(f"Booking hotel: {input.book_hotel_id}")

    await asyncio.sleep(1)

    if "buggy" in input.book_hotel_id.lower():
        # a simulated bug
        error = True
        if error:
            raise Exception("Error due to bug in code")

    return f"Booked hotel: {input.book_hotel_id}"


@activity.defn
async def book_flight(input: BookTripInput) -> str:
    activity.logger.info(f"Booking flight: {input.book_flight_id}")

    await asyncio.sleep(1)

    if "invalid" in input.book_flight_id.lower():
        # a business error, which cannot be retried
        raise ApplicationError(f"Flight {input.book_flight_id} is invalid", type="InvalidFlight", non_retryable=True)

    return f"Booked flight: {input.book_flight_id}"

@activity.defn
async def notify_user(input: BookTripInput) -> str:
    activity.logger.info(f"Notifying user: {input.book_user_id}")

    await asyncio.sleep(1)

    return f"Notified user: {input.book_user_id}"


@activity.defn
async def undo_book_car(input: BookTripInput) -> str:
    activity.logger.info(f"Undo car booking: {input.book_car_id}")

    await asyncio.sleep(1)

    return f"Unbooked car: {input.book_car_id}"


@activity.defn
async def undo_book_hotel(input: BookTripInput) -> str:
    activity.logger.info(f"Undo hotel booking: {input.book_hotel_id}")

    await asyncio.sleep(1)

    return f"Unbooked hotel: {input.book_hotel_id}"


@activity.defn
async def undo_book_flight(input: BookTripInput) -> str:
    activity.logger.info(f"Undo flight booking: {input.book_flight_id}")

    await asyncio.sleep(1)

    return f"Unbooked flight: {input.book_flight_id}"

File: go/app/activities.go
package app

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/temporal"
)

func BookFlight(ctx context.Context, flightId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Booking flight", "flightId", flightId)

	time.Sleep(1 * time.Second)

	if containsIgnoreCase(flightId, "flaky") {
		if activity.GetInfo(ctx).Attempt < 6 {
			return "", errors.New("flight booking service is currently unavailable")
		}
	}

	return fmt.Sprintf("Booked flight: %s", flightId), nil
}

func BookHotel(ctx context.Context, hotelId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Booking hotel", "hotelId", hotelId)

	time.Sleep(1 * time.Second)

	if containsIgnoreCase(hotelId, "buggy") {
		// a logical bug in the code
		error := true
		if error {
			return "", errors.New("error due to bug in code")
		}
	}

	return fmt.Sprintf("Booked hotel: %s", hotelId), nil
}

func BookCar(ctx context.Context, carId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Booking car", "carId", carId)

	time.Sleep(1 * time.Second)

	if containsIgnoreCase(carId, "invalid") {
		return "", temporal.NewNonRetryableApplicationError(
			fmt.Sprintf("Car '%s' is invalid", carId),
			"InvalidCar",
			nil,
		)
	}

	return fmt.Sprintf("Booked car: %s", carId), nil
}

func NotifyUser(ctx context.Context, userId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Notifying user", "userId", userId)

	time.Sleep(1 * time.Second)

	return fmt.Sprintf("Notified user: %s", userId), nil
}

func UndoBookFlight(ctx context.Context, flightId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Undo flight booking", "flightId", flightId)

	time.Sleep(1 * time.Second)

	return fmt.Sprintf("Unbooked flight: %s", flightId), nil
}

func UndoBookHotel(ctx context.Context, hotelId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Undo hotel booking", "hotelId", hotelId)

	time.Sleep(1 * time.Second)

	return fmt.Sprintf("Unbooked hotel: %s", hotelId), nil
}

func UndoBookCar(ctx context.Context, carId string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Undo car booking", "carId", carId)

	time.Sleep(1 * time.Second)

	return fmt.Sprintf("Unbooked car: %s", carId), nil
}

func containsIgnoreCase(s, substr string) bool {
	return strings.Contains(strings.ToLower(s), strings.ToLower(substr))
}
